### ISSUE TITLE
✨ Make secret name optional

### ISSUE THREAD

**Issue by sagikazarmark** [dagger/dagger#7358]
### What are you trying to do?

I have a function that creates a secret: https://daggerverse.dev/mod/github.com/sagikazarmark/daggerverse/registry-config@b45dbd7448bb967aca4a538af9ce7f042abf0316#RegistryConfig.secret

The current secret API takes a name at the moment: `dag.SetSecret("name", "value")`.

I'm not sure, but my guess is that name is supposed to be unique (ie. to make sure there is no interference between calls to the same function).

### Why is this important to you?

Right now I'm working around the issue mentioned above by exposing an optional secret name argument in the API. I'd rather not do that.

### How are you currently working around this?

I'm planning to generate a secret name based on the content: https://github.com/sagikazarmark/daggerverse/pull/104

That will make sure it's unique enough and won't require me to expose options in the API for the secret name.

**Comment by shykes** [comment]
FYI @jedevc @vito @helderco @aluzzardi I think we may need to revisit the design decisions that led to this point. Since then, we launched functions, modules, and didn't revisit the initial plan of adding "secrets providers". Is the current design still relevant? Do we still need the secret name?

**Comment by helderco** [comment]
One use case we have today is when adding secrets to a docker build. The secret name will be turned into the secret's `id` in the [Dockerfile](https://github.com/MikaelElkiaer/dagger-examples/blob/ae8a49103ea086f09eb50a423d4bed4871cb0c12/Dockerfile#L2):
```
RUN --mount=type=secret,id=secret-file curl
```

One problem today is that if you're passing a secret via the CLI, the secret name is an hash. So you currently need to work around that by either creating a new secret inside your function, with the plaintext of the `Secret` argument and the name that you use in the `Dockerfile`, or get the secret's name and pass it to the build as a build argument.

<details><summary>Build arg example</summary>

Fix for https://github.com/MikaelElkiaer/dagger-examples/blob/ae8a49103ea086f09eb50a423d4bed4871cb0c12/dockerfile-secrets/dagger/main.go:

```go
func (m *DockerfileSecrets) Build(ctx context.Context, buildContext *Directory, secretFile *Secret) (*Container, error) {
	secretName, err := secretFile.Name(ctx)
	if err != nil {
		return nil, err
	}
	return buildContext.
		DockerBuild(dagger.DirectoryDockerBuildOpts{
			BuildArgs: []BuildArg{
				{Name: "SECRET_ID", Value: secretName},
			},
			Secrets: []*Secret{secretFile},
		}), nil
}
```

</details>

In the docker build, you pass the secrets as `Secrets: []*Secret{...}`. If it were a map, we could use the map’s keys to be the IDs in the `Dockerfile`. But GraphQL doesn’t support maps so we’d need a new type with `name: String!` and `value: SecretID!` fields.

## Suggestion: add a `withName` field

However, since `Secret` already has a name, I think the simplest solution is adding a `withName` field to change the secret’s name. So the collapsed example above becomes much simpler:

```
func (m *DockerfileSecrets) Build(buildContext *Directory, secretFile *Secret) *Container {
	return buildContext.
		DockerBuild(dagger.DirectoryDockerBuildOpts{
			Secrets: []*Secret{secretFile.WithName("secret-file")},
		})
}
```

If you then **make the name in `SetSecret` optional** too (or create a new one without the name), then it won’t bother you if you don’t need to set it, but will also allow to lazily set it when needed. 

Not sure about the secret drives use case as I haven’t thought much about it. Curious to know what @sipsma has to say.

**Comment by helderco** [comment]
Forgot that we have a `dag.Secret("<name>")` and that secrets have been isolated in modules:
- https://github.com/dagger/dagger/pull/6601

\cc @jedevc

**Comment by jedevc** [comment]
Honestly, I'm not opposed to removing `dag.Secret`. It's incredibly fiddly as an API, and has global state, so would happily see it gone.

**Comment by shykes** [comment]
2 questions:

1. What happens if we remove the concept of secret name altogether? It seems like it may be getting in the way and not actually helping 

2. Do we still want a "pluggable secret provider" feature? If so what would that look like? The current design exists in part to keep the door open to such a design in the future. If we no longer feel the need for that, it's important to know. (cc @aluzzardi )

**Comment by helderco** [comment]
> 1. What happens if we remove the concept of secret name altogether? It seems like it may be getting in the way and not actually helping

How would it work in a docker build as explained in https://github.com/dagger/dagger/issues/7358#issuecomment-2110732668 ?

**Comment by helderco** [comment]
> 2. Do we still want a "pluggable secret provider" feature? If so what would that look like? The current design exists in part to keep the door open to such a design in the future. If we no longer feel the need for that, it's important to know.

I'm struggling to find a need for it when a normal module seems to do the trick.

**Comment by shykes** [comment]
@helderco yeah agreed. If the only reason for keeping a name is docker build... Better to solve the problem for docker build, and remove the secret name.

How does the docker CLI do it? Is it actually a build arg, or a distinct concept of build secret?

**Comment by TomChv** [comment]
> How does the docker CLI do it? Is it actually a build arg, or a distinct concept of build secret?

It's a different concept based on [the official doc](https://docs.docker.com/build/building/secrets/)

> [RUN](https://docs.docker.com/reference/dockerfile/#run) --mount=type=secret,id=mytoken \
    TOKEN=$(cat /run/secrets/mytoken) ...
>
> docker build --secret id=mytoken,src=$HOME/.aws/credentials .

Which has a unique identifier (`mytoken` in this example).

**Comment by jedevc** [comment]
Just had a thought on the some *implementation details* I though it would be worth writing down while I remembered.

Doing fully randomly generated secret names is not *quite* ideal, since this effectively acts as a cachebuster. Additionally, using the secret value is also not ideal, since the secret name might appear in different places in the output/etc.

However, there's another option - we can randomly generate secret names from the module+function id that it was created in. These are at least semi-stable, so won't bust cache everywhere (but still would on source changes, so not perfect).

Alternatively, if we're willing to remove the weird "ambient secret store" we have, which would break some code around using secrets in `DockerBuild`, we could potentially generate secret names at the point of usage in the LLB, with a hashed name which is *based on their usage* in the DAG - this would probably be the best caching experience we can get. Because this fully decouples secret names in dagger with secret names in the LLB, we can remove the names in dagger easily.

**Comment by shykes** [comment]
@jedevc I was wondering about that, but don't we want to *avoid* caching secrets anyway?

**Comment by jedevc** [comment]
> @jedevc I was wondering about that, but don't we want to _avoid_ caching secrets anyway?

This is debatable but yes! Generally, atm we just inherit buildkit's behavior - only the *name* of the secret is part of the cache key, the value isn't. So if you change the value, we keep caching, but if you change the name, then we can't.

I'd *like* to keep something close to this behavior, since otherwise, any step that uses a secret can never be cached (which would be a pretty sad regression as well as a substantial performance hit).

We already drift a little bit from this with secrets that are loaded in the CLI, where the name is *derived* from the value - I'm not generally a *huge* fan of that approach, since now a hash of the secret is technically in the cache key, but without the total refactor of the secret store I mention above to detach names entirely, I don't think we can avoid it.

**Comment by helderco** [comment]
I think there was a valid use case for making the value a part of the cache key. Otherwise, when a secret is used as an input to an operation, users may need to force a cache buster every time, to make sure the operation re-runs when the secret value changes. I think this was discussed with @sipsma, if you remember.

On the other hand, this isn't always the case and the secret's value shouldn't be the thing that busts the cache, just be available when the operation needs to rerun.

**Comment by helderco** [comment]
> Alternatively, if we're willing to remove the weird "ambient secret store" we have, which would break some code around using secrets in `DockerBuild`, we could potentially generate secret names at the point of usage in the LLB, with a hashed name which is _based on their usage_ in the DAG - this would probably be the best caching experience we can get. Because this fully decouples secret names in dagger with secret names in the LLB, we can remove the names in dagger easily.

I'll create an issue for an idea I discussed in GH comments. I think we need to deprecate `secrets: [Secret!]` in DockerBuild and introduce a new `secretArgs: [SecretArg!]` (similar to `buildArgs`). That forces providing a name for each secret in the context of the docker buil, since you need the secret "ID" in the Dockerfile (possibly fallback to the secret's name if the secret arg has an empty name, if there's a use case for not needing that name in the Dockerfile).

**Comment by jedevc** [comment]
Moved this out of v0.13.0, and into v0.14.0 - this still needs more experimentation, and the idea is to land v0.13.0 in next week, and this issue doesn't seem feasible in that timeframe.

**Comment by jedevc** [comment]
I'm de-milestoning this, instead of continuing to punt it.

I think our main goal is to deprecate `SetSecret` completely, which depends on the design discussions discussed in https://github.com/dagger/dagger/pull/8730

**Comment by shykes** [comment]
😭 we were so close to a small win... No opinion on the milestone part, but I do think we should keep trying for a quick win without freezing all improvements until we get a complete overhaul.
